Extend support for Sdk definition in .NET Core

mathbourgeois opened the pull request:
Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=58564 (I couldn't find an equivalent on the Github issues)

MsBuild has 3 methods of defining an SDK:
- Sdk attribute on the Project node
- Sdk node as a child of the Project node
- Sdk attribute on any Import. This permits explicit imports

Currently, MonoDevelop only supports the first one. Adding support for the other two cases, including not doing any implicit importing if the project already does so.

mathieubourgeois committed:
Adding support for all officially supported methods of declaring an Sdk usage in .Net Core
MonoDevelop only supported the Sdk attribute on the Project node, but there is also
- Sdk node as a child of the Project node. Like the Project attribute, the imports are then implicit
- Explicit Sdk imports (<Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />). In this case, the user already placed explicit imports so we shouldn't add them in its place

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/mono/monodevelop?pullRequest=5597) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/mono/monodevelop?pullRequest=5597) before we can accept your contribution.<br/><br/>:x: mathbourgeois [sign now](https://cla.opensource.microsoft.com/mono/monodevelop?pullRequest=5597)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/mono/monodevelop?pullRequest=5597) it.</sub>

mrward commented:
Maybe SdkWithImplicitImports could be replaced with a bool includeImplicitImports flag on the GetReferencedSdks method? We would need probably to store the sdk information and a flag whether it is an implicit import or not. I now understand why SdkWithImplicitImports was added here after reading the code again.

mathbourgeois commented:
For the tests, I entirely agree. I can add them to DotNetCoreProjectTests, it seems like the logical spot to add them. I'm not 100% sure about the coverage we should expect out of those tests though. Obviously, we should see if we can load a case with the Sdk node, and one with the explicit imports. Do we want to check for weirder cases, like a mix with different sdks? I'm not 100% sure how it should react to something like a Project Sdk attribute for an sdk, followed by explicit imports of those same sdk, so I'm not 100% sure where we should go.

For the properties, I agree that it's a bit messy. I agree with your solution for the GetReferencedSdks. I'll see how much I can reduce the complexity of the interface, it's indeed not very clear what each case does (at that point, settings the Sdk for the project should probably not be a property setter anyway, since it behaves now much more like a function, potentially removing nodes in the tree.

I'll try to work on this a little more this week, not sure when exactly I'll have the time though...

mrward commented:
Yeah I would just look at adding tests for loading a project with an Sdk element and a Import/@Sdk element. I would not be worried about some mix of things for now. We can check how Visual Studio on Windows reacts to some strange mix of sdk imports.

The test projects can be added in a subdirectory here - https://github.com/mono/monodevelop/tree/master/main/tests/test-projects

mathieubourgeois committed:
Cleaned Api based on Matt Ward's recommandations
- Renamed ProjectSdk back to Sdk to represent the Sdk set in the project node; to get a string representing all Sdk in the project, no matter where they're defined, call GetProjectSdksName()
- Removed SdkWithImplicitImports, call GetReferencedSdks and check for Implicit instead
- GetReferencedSdks now returns SdkDefinition, which knows if an Sdk has been defined implicitly or explicitly
- Provide a method to actually set the Sdk for the whole project, SetProjectSdk. This is to mirror the original behavior of Sdk, which had a setter. Like the original setter, it's unused. However, this function will Clear all sdks definition from the project, to ensure that the Sdk provided is the only one set. It will be set implicitly on the project node no matter what was before though...

mathieubourgeois committed:
Adding some basic load tests for Sdk formats, checking if the Project's sdk has been correctly defined and if the number of implicit imports matches how it's been defined in the project.

mathbourgeois commented:
I've cleaned up the api based on your recommandations and added some basic load tests. for all formats. Only thing I'm not 100% sure is the name to receive a string defining the sdk(s) for the whole project, no matter where they are defined : I named it GetProjectSdksName, which I don't think is perfect, but I couldn't find any better name off the top of my head. I'll be glad to change it if you have a better idea :smile:

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/MSBuildProject.cs:
```diff
@@ -983,17 +1106,41 @@ public XmlElement GetMonoDevelopProjectExtension (string section)
-			}
-			else
-				return Array.Empty<string> ();
+		public SdkDefinition[] GetReferencedSDKs ()
```
Unfortunately this is a public API break so we cannot change this to return SdkDefinitions. This would still need to return a string array.

I think GetReferencedSdks should return all referenced sdks and then we have a separate GetReferencedSdks method with a parameter that allows only the implicitly imported sdks to be returned, since the project model needs to import them itself.

mrward replied:
On the MSBuildProject think we can get away with the following new public methods:

    // Returns all sdks (explicit or implicit) - this is kind of obsolete due to the new method below.
    public string[] GetReferencedSdks ()

    public SdkDefinition[] GetSdks ()

mrward commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -78,7 +78,7 @@ protected override bool OnGetSupportsFramework (TargetFramework framework)
 		{
-			return project.MSBuildProject.Sdk != null;
+			return project.MSBuildProject.GetProjectSdksName() != null;
 		}
```
This could just use the GetReferencedSdks method.

mrward commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -406,7 +406,7 @@ static string GetDotNetCoreSdkRequiredBuildErrorMessage (bool isUnsupportedVersi
 		protected override void OnBeginLoad ()
 		{
-			dotNetCoreMSBuildProject.Sdk = Project.MSBuildProject.Sdk;
+			dotNetCoreMSBuildProject.Sdk = Project.MSBuildProject.GetProjectSdksName();
```
The dotNetCoreMSBuildProject.Sdk should still be the Project.MSBuildProject.Sdk value.

mrward commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -416,7 +416,7 @@ protected override void OnBeginLoad ()
 		protected bool IsWebProject (DotNetProject project)
 		{
-			return (project.MSBuildProject.Sdk?.IndexOf ("Microsoft.NET.Sdk.Web", System.StringComparison.OrdinalIgnoreCase) ?? -1) != -1;
+			return (project.MSBuildProject.GetProjectSdksName()?.IndexOf ("Microsoft.NET.Sdk.Web", System.StringComparison.OrdinalIgnoreCase) ?? -1) != -1;
```
I think this should change to just check the GetReferencedSdks instead of needing a new GetProjectSdksName method.

mrward commented on main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/DotNetCoreNuGetProject.cs:
```diff
@@ -98,7 +98,7 @@ public DotNetCoreNuGetProject (DotNetProject project)
 		{
-			return project.MSBuildProject.Sdk != null;
+			return project.MSBuildProject.GetProjectSdksName() != null;
 		}
```
This should just check the project has referenced sdks so it could just the GetReferencedSdks method instead of introducing a new method.

mrward commented on main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/DotNetProjectExtensions.cs:
```diff
@@ -171,7 +171,7 @@ public static IEnumerable<string> GetDotNetCoreTargetFrameworks (this Project pr
 		{
-			return project.MSBuildProject.Sdk != null;
+			return project.MSBuildProject.GetProjectSdksName() != null;
 		}
```
Same here.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/MSBuildProject.cs:
```diff
@@ -554,19 +561,135 @@ public string ToolsVersion
 		}
 
+		public class SdkDefinition
+		{
```
Not sure this need to be public if GetReferencedSdks will be returning a string array as before. If it needs to be public this should be in its own file.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/MSBuildProject.cs:
```diff
@@ -983,17 +1106,41 @@ public XmlElement GetMonoDevelopProjectExtension (string section)
+		/// Returns an Sdk attribute name defining all Sdks used in the project (implicit & explicit)
+		/// </summary>
+		public string GetProjectSdksName()
+		{
```
Not sure we need this. The information is available from GetReferencedSdks.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/MSBuildProject.cs:
```diff
@@ -983,17 +1106,41 @@ public XmlElement GetMonoDevelopProjectExtension (string section)
+		/// and will set an explicit sdk attribute on the project node.
+		/// </summary>
+		public void SetProjectSdk(string sdk)
+		{
```
I do not think we need this.

mrward commented on main/tests/Ide.Tests/MonoDevelop.Ide.Projects/DotNetCoreProjectTests.cs:
```diff
@@ -60,5 +60,38 @@ public async Task MoveResourceFilesToFolder ()
+				Assert.AreEqual (p.MSBuildProject.GetReferencedSDKs ().Where (x => x.Implicit).Count (), 0);
+			}
+		}
 	}
```
👍

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/MSBuildProject.cs:
```diff
@@ -554,19 +561,135 @@ public string ToolsVersion
+
+			public string QualifiedName {
+				get {
+					return !string.IsNullOrEmpty(Version) ? $"{Name}/{Version}" : Name;
```
Qualified name doesn't take into account that name and version can be null. Are we sure that nameAndVersion in the constructor _can_ be null?

mathieubourgeois committed:
Multiple extra cleanup :
- Put back GetReferencedSDKs to its already standard api
- Add a new version of GetReferencedSDKs to choose if we want implicit and/or explicit sdks
- Remove GetProjectSDKsName, replaced usage with one of the two GetReferencedSdks
- Correct DotNetCoreProjectExtension.OnBeginLoad to actually use Project Sdk instead of all referenced Sdks

mathieubourgeois committed:
Remove SetProjectSdk and ClearAllSdks; not necessary

mathieubourgeois committed:
Add missing check for sdk array generation
Could cause issue if we call GetReferencedSdks(bool, bool) before any calls to GetReferencedSdks(), returning null even if there were sdks defined
